### ISSUE TITLE
Transforms as an array

### ISSUE THREAD

**Issue by domoritz** [vega/vega-lite#2005]
[Dependencies of transforms](https://github.com/vega/vega-lite/wiki/Dependencies) are not obvious. Instead of implicitly encoding dependencies for the first explicit `transform`s (`filter`, `calculate`, `filterInvalid`), we want to make them an array. 

```js
transform: [{
  "filter": {"field": "a", "equal": 5}
}, {
  "filter": {"field": "a", "range": [0,5]}
}, {
  "filter": "datum.b > 10"
},{ // can have multiple filter"
  "filter": {"selection": "<selection-name>"}
}, {
  "calculate": "<formula>",
  "as": "<field_name>"
}]
```

In addition, we plan to add support for transforms that run before encoding. Note that you only need to do this if you want to do for example binning before a filter transform. The syntax for these transforms that are usually in encoding would be something like

```json
{
  "bin": true, "field": "acceleration", "as": "binned-acceleration"
}
```

```json
{
  "summarize": [
    {"aggregate": "sum", "field": "acceleration", "as": "sum_acceleration"},
    {"aggregate": "mean", "field": "horsepower", "as": "mean_horsepower"}
  ],
  "groupby": ["origin", "cylinders"]
}
```

this will be critical for statistical graphics that require post-aggregate transformations.

We can then express the crossfilter example from the paper without the weird "filterWith".

```js
{
  "data": {"url": "data/flights-2k.json"},
  "transform": [
    {"calculate": "hours(datum.date)", "as": "hour"}
    // {"timeUnit": "hours", "field": "date", "as": "hour"}
  ],
  "repeat": {"column": ["hour", "delay", "distance"]},
  "spec": {
    "transform": [
      // This binning process happens before any filtering
      {"bin": {"maxbins": 30}, "field": {"repeat": "column"}, "as": "binned-measure"}
    ],
    "layer": [
      {
        "selection": {
          "region": {
            "type": "interval",
            "project": {"channels": ["x"]},
            "resolve": "intersect"
          },
          "bins": {
            "type": "single",
            "project": {"channels": ["x"]},
            "on": "mousemove",
            "resolve": "union"
          }
        },
        "mark": "bar",
        "encoding": {
          "x": {"field": "binned-measure", "type": "Q"},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "steelblue"}
        }
      },{
        "transform": [
            // Instead of filterWith, selection filter becomes a Filter like EqualFilter and others.
            // Since this filter happens after the bin transform above, the bin uses the data pre-filter
            {"filter": {"selection": "bins || region"}} 
        ],
        "mark": "bar",
        "encoding": {
          "x": {"field": "binned-measure", "type": "Q"},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "goldenrod"}
        }
      }
    ]
  }
}
```

cc: @jheer @arvind

**Comment by arvind** [comment]
Not sure if this is a proposal, or something that has been decided? Apologies if it is the latter, but it'll take me a little bit to catch up and fully think through the implications. 

I'd like to dig into the motivation a little bit more. The dependency graph of transforms certainly looks complex on an initial glance, but does it overstate the complexity of the ordering as instantiated within the language? For example, `aggregate`, `bin`, and `timeUnit` can only be placed (I believe) within encoding channels, thus they must necessarily come _after_ "top-level" transforms. Similarly, the semantics of invoking these transforms within encoding channels dictates their ordering even though they can appear to be "siblings."

The one "special case" I remember is with the `filterWith` transform, introduced for selections. If I remember correctly, we had cases where we would sometimes want to interactively filter points before aggregation/binning, and other times after? But, it would seem to me that we should be able to infer that based on the selection definition? 

By asking these questions, I'm hoping to better understand that we're not making language-level changes in response to complexity that primarily exists in the compiler (though this can be a signal to improve the language).

My initial reaction to the actual proposal is to be concerned along the same lines as before: does this pull Vega-Lite closer to Vega, rather than keeping it true to itself? In particular:

(1) The array of transforms introduces a layer of indirection for operations that were previously more directly specified. We do have some precedent for this with the `calculate` transform, but that seems like a clearer need. 

(2) The implications on being able to systematically enumerate alternative designs. By moving closer to a Vega model, we grow our design space considerably. Whereas previously, we were restricted to aggregating/binning on data fields that are directly encoded (and not being able to filter, calculate, etc. on them), this proposal would now enable arbitrary orderings of them. Certainly we could impose such constraints within whatever higher-level enumerators we build, but they would no longer fall out cleanly as a property of the language design. 

Anyways, this is just my initial reaction. Will post more comments as I unpack it all in my head.

**Comment by domoritz** [comment]
Thank you for the detailed comment. I share your concerns but I also think that my initial description was a bit confusing. First, we are not removing aggregation or binning from the encoding. The only case where you would need to put aggregation or binning in the transforms array is when you want to filter or derive a field *after* binning or aggregation. We need this kind of functionality for crossfiltering and boxplots. Instead of adding special filters that run after binning (previously `filterWith`), we think a more explicit approach makes the semantics much clearer for users. 

Your concerns about enumeration make sense but we already need some constraints beyond what can be specified. I also don't think that the proposed syntax change allows for many non-obviously  wrong (read: not easy to enumerate correctly) specifications.

**Comment by kanitw** [comment]
In @domoritz's description above, maybe it's not clear that encoding's inline transform and implicit transform (e.g., filter non positive for log) still preserve the same order.  However, this enables us to 

1) do post-aggregation transform by using `aggregate` inside `transform`.  

2) it even provides stronger enforcement the transforms in encoding must necessarily come after "top-level" transforms.  Previously `filterWith` was a hack that violates this fact (comes after the transforms inside encoding).  With `transform` array in the cross filter example above, transformations in `transform` now always happen before transformations in `encoding`.

3) From our experience, the enforced dependencies between transformations in  `transform` (`calculate` and `filter`) are not always sensible.  As the authors of the system, we got tripped up multiple times ourselves (that formula has to come after filter or the other way round), so we won't be surprised if someone gonna get confused in the same way.   
Plus, it's going to be hard to extend transform in the future because it will just become harder to find an ordering that make sense.  

Instead of making things that has order an object/key-value pairs, we should just make them an array so the order (within the `transform` block) is explicit.  

> The array of transforms introduces a layer of indirection for operations that were previously more directly specified.

I'm not sure if this is truly an indirection since users still pretty much still specify the same thing except that the specification now have clear order of the transforms within `transform` block without having to resorting to look at the dependency chain.  

> The one "special case" I remember is with the filterWith transform, introduced for selections. If I remember correctly, we had cases where we would sometimes want to interactively filter points before aggregation/binning, and other times after? But, it would seem to me that we should be able to infer that based on the selection definition?

We may be able make the inference, but can't prevent the fact that it violates the rule that  the transforms in `encoding` must come after "top-level" transforms.

**Comment by arvind** [comment]
I'd like to propose a little smarter behavior for the placement of the `filter` transform when it's invoked with a selection. My motivation is in reaction to the example cross filter spec posted above. Namely, as proposed, adding cross filtering to a static repeated histogram is no longer an accretive process (i.e., just adding the selection and filter definitions) but would involve non-trivial restructuring of the specification.

I propose that if invoked with a selection that is projected over `encodings` (e.g., `"type": "interval", "encodings": ["x"]`) then the `filter` transform is **automatically** positioned to be evaluated _after_ encoding-level transforms. This makes semantic sense -- filtering with a projected selection imposes an order of transformations. If the selection is not projected over encodings, then `filter` occurs as it would otherwise.

Under this proposal, the cross filter spec moves closer to our original prototype, and specification of interaction techniques remains accretive:

```json
{
  "data": {"url": "data/flights-2k.json"},
  "transform": [
    {"calculate": "hours(datum.date)", "as": "hour"}
  ],
  "repeat": {"column": ["hour", "delay", "distance"]},
  "spec": {
    "layers": [
      {
        "selection": {
          "region": {
            "type": "interval", "encodings": ["x"]
          }
        },
        "mark": "bar",
        "encoding": {
          "x": {"field": {"repeat": "column"}, "type": "Q", "bin": {"maxbins": 30}},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "steelblue"}
        }
      },{
        "transform": [
          {"filter": {"selection": "region"}}
        ]
        "mark": "bar",
        "encoding": {
          "x": {"field": {"repeat": "column"}, "type": "Q", "bin":  {"maxbins": 30}},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "goldenrod"}
        }
      }
    ]
  }
}
```

Here, `filter` is using the `region` selection, which is projected over the `x` encoding, where the `bin` transform is defined. The only order that is meaningful is for `filter` to execute after `bin`, and should not necessitate extracting `bin` to the `transforms` array explicitly. 

Thus, `filter` acts like `filterWith`, but for more principled reasons. 

(I'll also note that projecting selections over encodings was introduced to account for the fact that it is difficult to address fields generated by encoding-level transforms).

**Comment by domoritz** [comment]
I think your proposal makes sense @arvind. How would a user override this behavior, though? Let's say I want to filter before binning.

**Comment by arvind** [comment]
* If filtering with an expression string (or field-projected selection), leave its position untouched.
* If filtering with an encoding-projected selection, move filter to occur _after_ encoding-level transforms. 

Good question about override -- perhaps we can introduce a config key to disable it? E.g., `"filterSelection": "manual"` or `"selectionFilter": "manual"` (where `"auto"` would be the default)?

**Comment by kanitw** [comment]
Accretive specification is an important point. But I don't know if I totally like what we try to do here yet (need to think more about it after the deadline).

For now, a few issues on the top of my mind:

1)  So if users specify 

```
transform: [
   <filter with encoding-projected-selection>,
   <calculate>,
   <filter>
]
```

do we also move things like you proposed or we keep the 1st one in this order?

2) config is only a top-level property, so I think that relying on top-level `config` to override this transform order to happen before / after encoding might be too distant from the part you try to affect.

**Comment by kanitw** [comment]
Having accretive specification process is important, but I think I don't like to have a case where the order in the transform can be violated. Before we get married to the idea that a transform can violate the order in the transform, I would like to throw another idea around. 

Currently `transform` always happen before inline transform inside encoding.  
Another alternative is to say that we can have the `transform` property to be either an array or an object the describe two arrays of transforms that get executed *before* and *after* the encoding inline transform.  This will still allow for accretive specification without the need to violate the order.

```
transform: Transform[] | {
   preEncoding: Transform[],
   postEncoding: Transform[]
}
```
Basically, users can still add `transform: {postEncoding: ...}` to the crossfilter example without the need to move all encoding inline transforms to the `transform` block.

(I don't like the key `preEncoding` and `postEncoding` that much yet, but also don't have a better alternative yet.)

**Comment by arvind** [comment]
Yeah, your proposal would make the order of operations completely unambiguous. But, I wonder if that's getting too low in the weeds for Vega-Lite? Will users know how/when to differentiate -- for example, what if they put a filter with an encoding-projected selection in the preEncoding block, that would get us back to our current situation. There's ambiguity throughout the remainder of the language, after all, so perhaps it's acceptable to introduce some here in a principled fashion. 

I'm thinking through two main cases:

1. Novice users/users who want to rapidly author specs. Vega-Lite's defaults are tuned for this type of user. This discussion highlights that the order of an encoding-selection-filter is non-trivial to think about, and is only semantically meaningful if it comes after encoding-level transforms. As a result, I think it would be acceptable to do the right for the user here. Changing the order of filter operations only affects (I think?) transforms that do some sort of aggregation/cumulative calculation. If users are explicitly invoking those as operators in a `transform` array and care about the order of operations, I would consider them:

2. Power users. We still support them by allowing them to override our defaults. Above, I suggested a `config` key, because this felt like the best place to override defaults. However, an alternative might be an inline key such as `"filter": {"selection": "region", "pinned": true}`. Thus, users who care about the particular ordering of transforms are still supported. 

One other reaction to explicitly separating pre/post-encoding transforms:

* What else could go in post-encoding? Technically none of the data fields generated by encoding-level transforms are addressable -- this was the motivation for introducing `encoding` projections for selections.

**Comment by kanitw** [comment]
Thanks @arvind.  I think the following point is very critical:

> none of the data fields generated by encoding-level transforms are addressable (expressible)

 I start to like the auto-moving filter proposal more especially the part that we auto-correct user's mistake. However, I think the proposal that we will allow pinning or overriding seems uncleaned to me. Instead, I think we might not need to allow overriding at all given the following rule:

>  we only re-order the filter with an encoding-projected selection, move `filter` to occur after encoding-level transforms.

Basically, there are two cases:

1) For encoding channel that has inline transform (`aggregate`, `bin`, `timeUnit`), we cannot apply the filter *before* the inline transform anyway so overriding does not really make sense.

2) For the encoding channel that does *not* have inline transform, there are two cases. 

  2.1) Raw plot -- then the filter can stay where it is declared. 

  2.2) Aggregate plot. -- The field in the channel is basically a group-by field in the aggregation. 
Applying the filter before and after the encoding inline transforms make the output equivalent but applying the filter earlier can make the aggregation faster. Thus we don't really need to move the filter in this case.  

Note that 2) will be rare since users can specify field directly in their selection, but we need to provide a good default anyway. 

Please let me know if there is any error in my analysis.  If there is none, I think it makes sense to automatically move the filter to after encoding inline transform when applicable (Case 1).   

We probably want to at least add info log whether the auto-move is triggered, so it's not too invisible. 
If the auto-moved filter is not the last item in the `transform` array, then we should throw a warning as the order in the transform array is violated.

**Comment by kanitw** [comment]
@jheer, @domoritz  and I chatted about this a bit today.
Plus, @jheer and @domoritz mentioned that we should not violate the transform order. 
In the meeting, we agreed that we can add `postTransform` property to the top-level object and keep the `transform` object the same.  

However, over lunch, @domoritz and I just realized that the correct behavior for the cross-filter example is applying the filter *after* binning but *before* aggregation, which makes things really tricky (and thus would make the `postTransform` proposal not really working).

@domoritz  and I think that given we will support the transform array for sure and we have a lot of other things to do for beta, we may modify the slide to have one additional step to move the bin to the `transform`.  (Unless we come up with a clean solution to this problem.)

**Comment by domoritz** [comment]
Another thing we realized over lunch is that if you want a post encoding transform, you need to reference the field through the channel instead of the field. take for example a spec where the same field is used binned and not binned. To disambiguate, which field (binned or not) a filter refers to, you need to specify the channel. However, what would the expression look like? One solution is to require `as` in the encoding but that is hardly discoverable.

**Comment by arvind** [comment]
* An encoding channel may include a `bin`, `timeUnit`, or `aggregate` transformation. 
* A selection may be projected over _multiple_ encoding channels that are transformed. 
* This selection is then be used to filter the input data. 
* There is only one order of operations that is semantically meaningful -- if the `filter` operation comes _after_ the last encoding-level transformation that the selection is projected over. 
* However, simply coming _after_ is not sufficient -- `aggregate` transforms may drop fields that are part of the original selection. Thus, the `filter` must come _immediately after_ the encoding-level transforms that the selection is projected over. 
* Requiring users to pull encoding-level transforms up to the top-level transforms array demands expertise and breaks accretive specification. Going from a static histogram to an interactive one, for example, would no longer just be about adding a selection + filter.
* Breaking accretive specification also hampers our ability to systematically enumerate the Vega-Lite design space. 
* Moreover, a selection may also be resolved to `others` (i.e., `union_others` or `intersect_others`). Thus, with multiple views, the filter cannot be pulled to the top-level but must happen within each subview. Specifying such a distinction explicitly is infeasible in Vega-Lite and, thus, the compiler must analyze selection definitions to generate and appropriately position the filter operator.
* Having a `filter` operation _within_ the `transform` array that moves around, as I previously proposed, is a non-starter as it sharply breaks a user's ability to reason about the order of operations and would require additional definitions to simply maintain specification (array) order.
* Adding an encoding-level `filter` transform for selections is also not ideal since a selection may be projected over multiple encoding fields that are transformed.
* Instead, I now propose a `selectionFilter` key at the top-level of a unit specification which would take a string of selection names that are logically composed. For instance, `"selectionFilter": "region && grid"`. The `region` and `grid` selections would be inspected and, if they were projected over transformed encoding channels, the filter operator would be inserted with the necessary ordering (i.e., after binning/timeUnit, or aggregate, or within subviews if resolved as such).

More concretely, consider the following setup for layered cross filtering:

```json
{
  "data": {"url": "data/flights-2k.json"},
  "transform": [
    {"calculate": "hours(datum.date)", "as": "hour"}
  ],
  "repeat": {"column": ["hour", "delay", "distance"]},
  "spec": {
    "layers": [
      {
        "selection": {
          "region": {
            "type": "interval", "encodings": ["x"],
            "resolve": "intersect"
          }
        },
        "mark": "bar",
        "encoding": {
          "x": {"field": {"repeat": "column"}, "type": "Q", "bin": {"maxbins": 30}},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "steelblue"}
        }
      },{
        "selectionFilter": "region",
        "mark": "bar",
        "encoding": {
          "x": {"field": {"repeat": "column"}, "type": "Q", "bin":  {"maxbins": 30}},
          "y": {"aggregate": "count", "field": "*", "type": "quantitative"},
          "color": {"value": "goldenrod"}
        }
      }
    ]
  }
}
```

Here, the `region` interval selection is projected over the `x` encoding channel, which is binned. Each concatenated view is driven by a separate aggregated data source, which only includes one (of the three) binned fields. Thus, the binning operators must be pulled to the top-level/source dataset. As the selection is resolved to `intersect`, the `filter` operator could come next at the top-level, and then finally the necessary `aggregate` operators in derived data sources for each view.

Had the selection be defined to resolve to `intersect_others` (or `union_others`), the `filter` operator would need to be placed ahead of the `aggregate` operators in each of the derived data sources. 

We cannot fix the `selectionFilter` operator to always appear after bin/timeUnit because a selection may be projected over an aggregated encoding. In single views, this is fairly straightforward -- the filter should appear after the aggregate. Multiple views will require some notion of a "derived selection" to re-project over a non-aggregated field (e.g., "categories with mean above X") that is likely best supported in a subsequent release.

**Comment by jheer** [comment]
Thanks for laying this our @arvind. I have a syntactical question related to this proposal: Why use a new form of expression (e.g., `sel1 && sel2`). It seems a JSON syntax might be reasonable (e.g., not too verbose) and easier to manipulate and enumerate. Perhaps using a similar syntax as the JSON filtering syntax @kanitw developed, but for selections? (Might be nice if we can be consistent!)

For example:
```js
"selectionFilter": "sel" // single selection
"selectionFilter": {"not": "sel1"} // negation
"selectionFilter": {"and": ["sel1", "sel2"]} // conjunction
"selectionFilter": {"or": ["sel1", "sel2"]} // disjunction
```

**Comment by arvind** [comment]
Oh, absolutely. I had totally forgotten that syntax was already available. Perhaps we should adopt it for the `conditional` statements as well (though only `not` is currently supported)? Similarly here, perhaps it's worth initially restricting the `selectionFilter` to simply `sel` and `{"not": "sel"}`.

**Comment by kanitw** [comment]
We already have transform array and figure out how to deal with bin without a special `selectionFilter` key.  The only task left is adding bin, aggregate, timeUnit to `transform` so we can calculate/filter based on these values (https://github.com/vega/vega-lite/issues/2344) -- so let's close this.
